controller/volume_attachment_controller.go

```diff
@@ -0,0 +1,413 @@
+	vac.logger.WithError(err).Warnf("Error syncing Longhorn VolumeAttachment %v", key)
+	vac.queue.AddRateLimited(key)
+	return
+}
+
+func (vac *VolumeAttachmentController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: failed to sync VolumeAttachment %v", vac.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != vac.namespace {
+		return nil
+	}
+	return vac.reconcile(name)
+}
+
+func (vac *VolumeAttachmentController) reconcile(vaName string) (err error) {
+	va, err := vac.ds.GetLHVolumeAttachment(vaName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	isResponsible, err := vac.isResponsibleFor(va)
+	if err != nil {
+		return err
+	}
+	if !isResponsible {
+		return nil
+	}
+
+	//log := getLoggerForLHVolumeAttachment(vac.logger, va)
+
+	vol, err := vac.ds.GetVolume(va.Spec.Volume)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// remove finalizer
+			return
+		}
+		return err
+	}
+
+	// handle VA deletion flow
+
+	existingVA := va.DeepCopy()
+	defer func() {
+		if err != nil {
+			return
+		}
+		if reflect.DeepEqual(existingVA.Status, va.Status) {
+			return
+		}
+		_, err = vac.ds.UpdateLHVolumeAttachmetStatus(va)
+		return
+	}()
+
+	vol, err = vac.handleVolumeDetachment(va, vol)
+	if err != nil {
+		return err
+	}
+
+	vol, err = vac.handleVolumeAttachment(va, vol)
+	if err != nil {
+		return err
+	}
+
+	return vac.handleVAStatusUpdate(va, vol)
+}
+
+func (vac *VolumeAttachmentController) handleVolumeDetachment(va *longhorn.VolumeAttachment, vol *longhorn.Volume) (*longhorn.Volume, error) {
+	// Volume is already trying to detach
+	if vol.Spec.NodeID == "" {
+		return vol, nil
+	}
+
+	for _, attachment := range va.Spec.Attachments {
+		// Found one attachment that is still requesting volume to attach to the current node
+		if attachment.NodeID == vol.Spec.NodeID && verifyAttachmentParameters(attachment, vol) {
+			return vol, nil
+		}
+	}
+
+	// There is no attachment that request the current vol.Spec.NodeID.
+	// Therefore, set desire state of volume to empty
+	vol.Spec.NodeID = ""
+	// reset the attachment parameter
+	setAttachmentParameter(&longhorn.Attachment{}, vol)
+	return vac.ds.UpdateVolume(vol)
+}
+
+func (vac *VolumeAttachmentController) handleVolumeAttachment(va *longhorn.VolumeAttachment, vol *longhorn.Volume) (*longhorn.Volume, error) {
+	// Wait for volume to be ready to be fully detached
+	if vol.Spec.NodeID != "" ||
+		vol.Spec.MigrationNodeID != "" ||
+		vol.Status.PendingNodeID != "" ||
+		vol.Status.State != longhorn.VolumeStateDetached {
+		return vol, nil
+	}
+
+	attachment := selectAttachmentToAttach(va)
+	if attachment == nil {
+		return vol, nil
+	}
+
+	vol.Spec.NodeID = attachment.NodeID
+	setAttachmentParameter(attachment, vol)
+	return vac.ds.UpdateVolume(vol)
+}
+
+func selectAttachmentToAttach(va *longhorn.VolumeAttachment) *longhorn.Attachment {
```

Comment:
Why do we care about the priority? instead of determining it by request time?

Reply:
We want to be able to interrupt the attachment ticket based on priority. For example. a backup AD ticket is running. A csi-attacher AD ticket come in, we want to abort the backup to prioritize the workload. This logic is still under implementation